src/circuit/auto-relay.js

```diff
@@ -0,0 +1,170 @@
+'use strict'
+
+const debug = require('debug')
+const log = debug('libp2p:auto-relay')
+log.error = debug('libp2p:auto-relay:error')
+
+const multiaddr = require('multiaddr')
+const MulticodecTopology = require('libp2p-interfaces/src/topology/multicodec-topology')
+
+const { relay: multicodec } = require('./multicodec')
+const { canHop } = require('./circuit/hop')
+
+class AutoRelay {
+  /**
+   * Creates an instance of AutoRelay
+   * @constructor
+   * @param {object} params
+   * @param {Libp2p} params.libp2p
+   * @param {number} params.maxListeners maximum number of relays to listen.
+   */
+  constructor ({ libp2p, maxListeners }) {
+    this._libp2p = libp2p
+    this._peerId = libp2p.peerId
+    this._registrar = libp2p.registrar
+    this._peerStore = libp2p.peerStore
+    this._connectionManager = libp2p.connectionManager
+    this._transportManager = libp2p.transportManager
+
+    this.maxListeners = maxListeners
+
+    /**
+     * @type {Set<string>}
+     */
+    this._listenRelays = new Set()
+
+    this._onPeerConnected = this._onPeerConnected.bind(this)
+    this._onPeerDisconnected = this._onPeerDisconnected.bind(this)
+
+    // register protocol with topology
+    const topology = new MulticodecTopology({
+      multicodecs: multicodec,
+      handlers: {
+        onConnect: this._onPeerConnected,
+        onDisconnect: this._onPeerDisconnected
+      }
+    })
+    this._registrar.register(topology)
+
+    // TODO: should proactively try to connect?
+    // We need to figure out a more general approach for the autoDial with topologies
+  }
+
+  /**
+   * Registrar notifies a connection successfully with circuit protocol.
+   * @private
+   * @param {PeerId} peerId remote peer-id
+   * @param {Connection} connection connection to the peer
+   */
+  async _onPeerConnected (peerId, connection) {
+    // Check if already listening on enough relays
+    if (this._listenRelays.size >= this.maxListeners) {
+      return
+    }
```

Comment:
This is not really taking advantage of the topology, and we might not want to use multicodec for this. Right now, every peer that supports the relay protocol is going to get added as a peer to the topology, which isn't good because nearly all peers support the relay protocol (both go and js default to this for STOP support).

We still want to listen on protocol updates to see who supports relay, and then check `canHop` on those. If they do support HOP we should add them to a list of known HOP relays. This could be as simple as adding metadata to the peer store (hop_relay=true or something) so we can persist this knowledge. I think separating the check from the listen would be good. This would enable us to more easily check for more relays without listening on them. For example, I really only need 1 relay to listen on, but Ideally I'd like to know about a couple in case the first has issues.

> Should we keep a backup set of connected peers that have the relay protocol, so that we don't need to iterate all the connections when a peer disconnects?

I think tracking the metadata would resolve this. We could ditch the topology altogether for now and just listen on the peerstore protocol changes for the checks. I'd definitely try to keep this as simple as possible for now.

Reply:
I will switch to listen on protocol event changes. The metadata seems nice to have this information for persistence on restart. But, (as mentioned earlier in a PR review), it feels strange to me to mix the metadata that the application layer might insert and libp2p metadata. It would be nice to think of a way of at least flagging the origin/context of the metadata insertion